File: lib/data/updates/update_manifest.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart' as hashes;
import 'package:pointycastle/api.dart';
import 'package:pointycastle/asymmetric/api.dart';
import 'package:pointycastle/digests/sha256.dart';
import 'package:pointycastle/signers/rsa_signer.dart';

import '../../domain/models/release_identity.dart';

const updateApplicationId = 'dev.shiori.reader';
const updateSignatureAlgorithm = 'rsa3072-pkcs1-sha256';
const maxUpdateManifestBytes = 4 * 1024 * 1024;

/// Trust is supplied by the installed app, never by a downloaded manifest.
final class UpdatePublicKey {
  UpdatePublicKey.fromJson(Map<String, dynamic> json)
    : modulusHex = json['modulus'] as String,
      keyId = json['keyId'] as String {
    if (json['algorithm'] != updateSignatureAlgorithm ||
        json['exponent'] != 65537 ||
        !RegExp(r'^[89a-f][0-9a-f]{767}$').hasMatch(modulusHex) ||
        BigInt.parse(modulusHex, radix: 16).isEven ||
        keyId != hashes.sha256.convert(_hexBytes(modulusHex)).toString()) {
      throw const FormatException('Invalid update public key');
    }
  }

  final String modulusHex;
  final String keyId;

  bool verify(Uint8List bytes, Uint8List signature) {
    if (bytes.isEmpty ||
        bytes.length > maxUpdateManifestBytes ||
        signature.length != 384) {
      return false;
    }
    final signer = RSASigner(SHA256Digest(), '0609608648016503040201')
      ..init(
        false,
        PublicKeyParameter<RSAPublicKey>(
          RSAPublicKey(BigInt.parse(modulusHex, radix: 16), BigInt.from(65537)),
        ),
      );
    try {
      return signer.verifySignature(bytes, RSASignature(signature));
    } on ArgumentError {
      return false;
    }
  }
}

Uint8List _hexBytes(String hex) => Uint8List.fromList([
  for (var i = 0; i < hex.length; i += 2)
    int.parse(hex.substring(i, i + 2), radix: 16),
]);

ReleaseIdentity _identity(Map<String, dynamic> json) {
  final result = ReleaseIdentity(
    tag: json['tag'] as String,
    version: json['version'] as String,
    build: json['build'] as int,
    commit: json['commit'] as String,
  );
  if (json['applicationId'] != updateApplicationId ||
      json['channel'] != result.channel.name) {
    throw const FormatException('Release identity mismatch');
  }
  return result;
}

/// Read assets/release/build-info.json; local/iOS builds are marked development.
final class BundledRelease {
  BundledRelease._(this.identity, this.publicKey);
  final ReleaseIdentity identity;
  final UpdatePublicKey publicKey;

  static BundledRelease? fromJson(Map<String, dynamic> json) {
    if (json['schemaVersion'] != 1) {
      throw const FormatException('Unknown build identity');
    }
    if (json['development'] == true) return null;
    return BundledRelease._(
      _identity(json['identity'] as Map<String, dynamic>),
      UpdatePublicKey.fromJson(json['publicKey'] as Map<String, dynamic>),
    );
  }

  void verifyInstalledVersion(String version, int build) {
    if (identity.version != version || identity.build != build) {
      throw const FormatException(
        'Installed version differs from bundled identity',
      );
    }
  }
}

final class UpdateFile {
  const UpdateFile(this.path, this.size, this.sha256);
  final String path;
  final int size;
  final String sha256;
}

final class UpdateAsset {
  UpdateAsset._(
    this.platform,
    this.name,
    this.size,
    this.sha256,
    this.minimumSystem,
    this.certificateSha256,
    this.updaterProtocol,
    List<UpdateFile> files,
  ) : files = List.unmodifiable(files);
  final String platform;
  final String name;
  final int size;
  final String sha256;
  final String minimumSystem;
  final String? certificateSha256;
  final int? updaterProtocol;
  final List<UpdateFile> files;
}

final class UpdateManifest {
  UpdateManifest._(this.identity, List<UpdateAsset> assets)
    : assets = List.unmodifiable(assets);
  final ReleaseIdentity identity;
  final List<UpdateAsset> assets;

  static UpdateManifest verifyAndRead(
    Uint8List bytes,
    Uint8List signature,
    UpdatePublicKey trustedKey, {
    required String expectedTag,
  }) {
    if (!trustedKey.verify(bytes, signature)) {
      throw const FormatException('Invalid update manifest signature');
    }
    final json = jsonDecode(utf8.decode(bytes)) as Map<String, dynamic>;
    final identity = _identity(json);
    if (json['schemaVersion'] != 1 ||
        json['keyId'] != trustedKey.keyId ||
        json['algorithm'] != updateSignatureAlgorithm ||
        identity.tag != expectedTag) {
      throw const FormatException('Unsupported update manifest');
    }
    final assets = <UpdateAsset>[];
    final platforms = <String>{};
    for (final item in json['assets'] as List<dynamic>) {
      final asset = item as Map<String, dynamic>;
      final platform = asset['platform'] as String;
      if (!['android', 'windows-x64'].contains(platform) ||
          !platforms.add(platform)) {
        throw const FormatException('Invalid update platform');
      }
      final suffix = platform == 'android' ? 'android.apk' : 'windows-x64.zip';
      final name = asset['name'] as String;
      final minimum = asset['minimumSystem'] as String;
      if (name != 'shiori-reader-${identity.tag}-$suffix' || minimum.isEmpty) {
        throw const FormatException('Invalid update asset');
      }
      final size = _size(asset['size'], allowZero: false);
      final digest = _digest(asset['sha256']);
      final files = <UpdateFile>[];
      String? certificate;
      int? protocol;
      if (platform == 'android') {
        certificate = _digest(asset['certificateSha256']);
      } else {
        protocol = asset['updaterProtocol'] as int;
        final entries = asset['files'] as List<dynamic>;
        if (protocol != 1 || entries.isEmpty || entries.length > 10000) {
          throw const FormatException(
            'Unsupported updater protocol or file count',
          );
        }
        final names = <String>{};
        var total = 0;
        for (final raw in entries) {
          final file = raw as Map<String, dynamic>;
          final path = file['path'] as String;
          if (!isSafeUpdatePath(path) || !names.add(path.toLowerCase())) {
            throw const FormatException('Invalid update file path');
          }
          final length = _size(file['size'], allowZero: true);
          total += length;
          if (total > 2 * 1024 * 1024 * 1024) {
            throw const FormatException('Update too large');
          }
          files.add(UpdateFile(path, length, _digest(file['sha256'])));
        }
        for (final file in files) {
          final parts = file.path.toLowerCase().split('/');
          for (var i = 1; i < parts.length; i++) {
            if (names.contains(parts.take(i).join('/'))) {
              throw const FormatException('File and directory collision');
            }
          }
        }
        if (!names.contains('shiori.exe') || !names.contains('data/app.so')) {
          throw const FormatException('Missing Windows runtime');
        }
      }
      assets.add(
        UpdateAsset._(
          platform,
          name,
          size,
          digest,
          minimum,
          certificate,
          protocol,
          files,
        ),
      );
    }
    if (platforms.length != 2) {
      throw const FormatException('Incomplete platform assets');
    }
    return UpdateManifest._(identity, assets);
  }
}

int _size(dynamic value, {required bool allowZero}) {
  if (value is! int ||
      value < (allowZero ? 0 : 1) ||
      value > 2 * 1024 * 1024 * 1024) {
    throw const FormatException('Invalid update size');
  }
  return value;
}

String _digest(dynamic value) {
  if (value is! String || !RegExp(r'^[0-9a-f]{64}$').hasMatch(value)) {
    throw const FormatException('Invalid SHA-256');
  }
  return value;
}

bool isSafeUpdatePath(String path) {
  if (path.isEmpty ||
      path.length > 1024 ||
      RegExp(r'[\\:\x00-\x1f<>"|?*]').hasMatch(path)) {
    return false;
  }
  return path
      .split('/')
      .every(
        (part) =>
            part.isNotEmpty &&
            part != '.' &&
            part != '..' &&
            !part.endsWith('.') &&
            !part.endsWith(' ') &&
            !RegExp(
              r'^(con|prn|aux|nul|com[1-9]|lpt[1-9])(?:\.|$)',
              caseSensitive: false,
            ).hasMatch(part),
      );
}

File: tool/update_manifest.py
"""Signed update protocol. OpenSSL handles RSA; private keys never appear in output."""

import argparse
import base64
import hashlib
import json
import os
import re
import stat
import subprocess
import tempfile
import zipfile
from pathlib import Path

from release_android import ReleaseCheckError, version

ALGORITHM = "rsa3072-pkcs1-sha256"
APPLICATION_ID = "dev.shiori.reader"
MAX_SIZE = 2 * 1024 * 1024 * 1024
MAX_MANIFEST = 4 * 1024 * 1024
IDENTITY_ASSET = "assets/release/build-info.json"


def openssl(*args, input_bytes=None):
    result = subprocess.run(
        [os.environ.get("OPENSSL", "openssl"), *map(str, args)],
        input=input_bytes,
        capture_output=True,
        check=False,
    )
    if result.returncode:
        raise ReleaseCheckError("OpenSSL update signing operation failed")
    return result.stdout


def encoded(value):
    return (
        json.dumps(value, ensure_ascii=True, sort_keys=True, separators=(",", ":"))
        + "\n"
    ).encode()


def read_json(raw):
    def pairs(items):
        value = {}
        for key, entry in items:
            if key in value:
                raise ReleaseCheckError("Duplicate JSON field")
            value[key] = entry
        return value

    return json.loads(raw, object_pairs_hook=pairs)


def public_key(value):
    modulus = value.get("modulus", "")
    if (
        value.get("algorithm") != ALGORITHM
        or value.get("exponent") != 65537
        or not isinstance(modulus, str)
        or not re.fullmatch("[89a-f][0-9a-f]{767}", modulus)
        or int(modulus, 16) % 2 != 1
        or value.get("keyId") != hashlib.sha256(bytes.fromhex(modulus)).hexdigest()
    ):
        raise ReleaseCheckError("Invalid update public key")
    return value


def key_from_private(path):
    details = openssl("pkey", "-in", path, "-text_pub", "-noout").decode()
    if not re.search(r"(?:publicExponent|Exponent): 65537 \(0x10001\)", details):
        raise ReleaseCheckError("Update key exponent must be 65537")
    output = openssl("rsa", "-in", path, "-modulus", "-noout").decode().strip()
    modulus = output.removeprefix("Modulus=").lower()
    return public_key(
        {
            "algorithm": ALGORITHM,
            "exponent": 65537,
            "modulus": modulus,
            "keyId": hashlib.sha256(bytes.fromhex(modulus)).hexdigest(),
        }
    )


def identity(pubspec, tag, commit):
    name, build = version(pubspec, tag)
    if not re.fullmatch(
        r"v(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-beta\.[1-9]\d*)?", tag
    ):
        raise ReleaseCheckError("Unsupported update release tag")
    if not re.fullmatch("[0-9a-f]{40}", commit):
        raise ReleaseCheckError("Update identity requires a full commit SHA")
    return {
        "applicationId": APPLICATION_ID,
        "tag": tag,
        "version": name,
        "build": int(build),
        "channel": "beta" if "-beta." in tag else "stable",
        "commit": commit,
    }


def check_bundled(raw, expected, key):
    bundled = read_json(raw)
    if bundled != {"schemaVersion": 1, "identity": expected, "publicKey": key}:
        raise ReleaseCheckError("Package build identity or trusted key mismatch")


def safe_path(path):
    if not path or len(path) > 1024 or re.search(r'[\\:\x00-\x1f<>"|?*]', path):
        return False
    return all(
        part
        and part not in (".", "..")
        and not part.endswith((".", " "))
        and not re.match(
            r"^(con|prn|aux|nul|com[1-9]|lpt[1-9])(?:\.|$)", part, re.IGNORECASE
        )
        for part in path.split("/")
    )


def sha(path):
    with path.open("rb") as source:
        return hashlib.file_digest(source, "sha256").hexdigest()


def windows_files(path, expected, key):
    files, names, total = [], set(), 0
    with zipfile.ZipFile(path) as package:
        entries = package.infolist()
        if not entries or len(entries) > 10000:
            raise ReleaseCheckError("Invalid Windows update file count")
        for entry in entries:
            name = entry.filename
            if (
                not safe_path(name)
                or name.lower() in names
                or entry.is_dir()
                or stat.S_ISLNK(entry.external_attr >> 16)
                or entry.flag_bits & 1
            ):
                raise ReleaseCheckError("Unsafe Windows update entry")
            names.add(name.lower())
            total += entry.file_size
            if total > MAX_SIZE:
                raise ReleaseCheckError("Windows update exceeds size limit")
            with package.open(entry) as source:
                digest = hashlib.file_digest(source, "sha256").hexdigest()
            files.append({"path": name, "size": entry.file_size, "sha256": digest})
        for name in names:
            parts = name.split("/")
            if any("/".join(parts[:i]) in names for i in range(1, len(parts))):
                raise ReleaseCheckError("Windows file/directory collision")
        if not {"shiori.exe", "data/app.so"}.issubset(names):
            raise ReleaseCheckError("Missing Windows runtime files")
        bundled = package.getinfo("data/flutter_assets/" + IDENTITY_ASSET)
        if bundled.file_size > 16384:
            raise ReleaseCheckError("Build identity too large")
        check_bundled(package.read(bundled), expected, key)
    return sorted(files, key=lambda item: item["path"])


def create_manifest(root, assets, tag, commit, key):
    expected = identity(
        (root / "pubspec.yaml").read_text(encoding="utf-8"), tag, commit
    )
    key = public_key(key)
    result = dict(
        schemaVersion=1, algorithm=ALGORITHM, keyId=key["keyId"], **expected, assets=[]
    )
    for platform, suffix, metadata_name in [
        ("android", "android.apk", "release-info.json"),
        ("windows-x64", "windows-x64.zip", "release-info-windows-x64.json"),
    ]:
        metadata = read_json((assets / metadata_name).read_bytes())
        name = f"shiori-reader-{tag}-{suffix}"
        path = assets / name
        if (
            metadata["tag"] != tag
            or metadata["commit"] != commit
            or metadata["versionName"] != expected["version"]
            or metadata["versionCode"] != expected["build"]
            or metadata.get("apk" if platform == "android" else "archive") != name
            or path.is_symlink()
            or not 0 < path.stat().st_size <= MAX_SIZE
            or metadata["sha256"] != sha(path)
        ):
            raise ReleaseCheckError("Update asset metadata mismatch")
        asset = {
            "platform": platform,
            "name": name,
            "size": path.stat().st_size,
            "sha256": metadata["sha256"],
        }
        if platform == "android":
            certificate = metadata["certificateSha256"]
            if metadata["applicationId"] != APPLICATION_ID or not re.fullmatch(
                "[0-9a-f]{64}", certificate
            ):
                raise ReleaseCheckError("Android package identity mismatch")
            with zipfile.ZipFile(path) as apk:
                bundled = apk.getinfo("assets/flutter_assets/" + IDENTITY_ASSET)
                if bundled.file_size > 16384:
                    raise ReleaseCheckError("Build identity too large")
                check_bundled(apk.read(bundled), expected, key)
            asset.update(minimumSystem="24", certificateSha256=certificate)
        else:
            if metadata["platform"] != "windows-x64":
                raise ReleaseCheckError("Windows platform identity mismatch")
            asset.update(
                minimumSystem="10.0.17763",
                updaterProtocol=1,
                files=windows_files(path, expected, key),
            )
        result["assets"].append(asset)
    raw = encoded(result)
    if len(raw) > MAX_MANIFEST:
        raise ReleaseCheckError("Update manifest too large")
    return raw


def sign_manifest(raw, private_pem, key):
    with tempfile.TemporaryDirectory(prefix="shiori-update-sign-") as directory:
        private = Path(directory) / "private.pem"
        private.write_bytes(private_pem)
        private.chmod(0o600)
        if key_from_private(private) != public_key(key):
            raise ReleaseCheckError("Update signing key does not match bundled trust")
        signature = openssl("dgst", "-sha256", "-sign", private, input_bytes=raw)
        if len(signature) != 384:
            raise ReleaseCheckError("Invalid update signature size")
        return signature


def verify_signature(raw, signature, key):
    key = public_key(key)
    if not 0 < len(raw) <= MAX_MANIFEST or len(signature) != 384:
        raise ReleaseCheckError("Invalid update signature input")
    with tempfile.TemporaryDirectory(prefix="shiori-update-verify-") as directory:
        directory = Path(directory)
        # PKCS#1 RSAPublicKey DER, fixed 3072-bit positive modulus and exponent 65537.
        der = bytes.fromhex("3082018a0282018100" + key["modulus"] + "0203010001")
        (directory / "public.der").write_bytes(der)
        public = openssl(
            "rsa",
            "-RSAPublicKey_in",
            "-inform",
            "DER",
            "-in",
            directory / "public.der",
            "-pubout",
        )
        (directory / "public.pem").write_bytes(public)
        (directory / "manifest.sig").write_bytes(signature)
        openssl(
            "dgst",
            "-sha256",
            "-verify",
            directory / "public.pem",
            "-signature",
            directory / "manifest.sig",
            input_bytes=raw,
        )


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("command", choices=["identity", "manifest", "keygen"])
    parser.add_argument("--root", type=Path, default=Path.cwd())
    parser.add_argument("--assets", type=Path, default=Path("build/release-assets"))
    parser.add_argument("--tag", default=os.environ.get("GITHUB_REF_NAME", ""))
    parser.add_argument("--commit", default=os.environ.get("GITHUB_SHA", ""))
    parser.add_argument("--private-key", type=Path)
    parser.add_argument("--public-key", type=Path)
    args = parser.parse_args()
    if args.command == "keygen":
        if (
            not args.private_key
            or not args.public_key
            or args.private_key.exists()
            or args.public_key.exists()
        ):
            raise ReleaseCheckError("Supply new private/public key output paths")
        # Generate in memory; exclusive creation prevents overwriting an existing key.
        secret = openssl(
            "genpkey", "-algorithm", "RSA", "-pkeyopt", "rsa_keygen_bits:3072"
        )
        with args.private_key.open("xb") as output:
            args.private_key.chmod(0o600)
            output.write(secret)
        key = key_from_private(args.private_key)
        with args.public_key.open("xb") as output:
            output.write(encoded(key))
        print(
            "Update key pair created. Store the private key securely; publish only the public key."
        )
        return
    key = public_key(read_json(os.environ.get("UPDATE_PUBLIC_KEY_JSON", "{}")))
    if args.command == "identity":
        expected = identity(
            (args.root / "pubspec.yaml").read_text(encoding="utf-8"),
            args.tag,
            args.commit,
        )
        (args.root / IDENTITY_ASSET).write_bytes(
            encoded({"schemaVersion": 1, "identity": expected, "publicKey": key})
        )
    else:
        raw = create_manifest(args.root, args.assets, args.tag, args.commit, key)
        secret = base64.b64decode(
            os.environ.get("UPDATE_SIGNING_KEY_PEM_B64", ""), validate=True
        )
        if not secret:
            raise ReleaseCheckError("Missing UPDATE_SIGNING_KEY_PEM_B64")
        signature = sign_manifest(raw, secret, key)
        (args.assets / "update-manifest.json").write_bytes(raw)
        (args.assets / "update-manifest.sig").write_bytes(signature)


if __name__ == "__main__":
    try:
        main()
    except (ValueError, OSError, KeyError, TypeError, zipfile.BadZipFile):
        raise SystemExit(
            "Update preparation failed: check release identity, assets and signing configuration."
        )
